### ISSUE TITLE
Passing Network to mpi_child instead of Params causes intermittent crashes

### ISSUE THREAD

**Issue by blakecaldwell** [jonescompneurolab/hnn-core#285]
I would notice crashes every 3-5 simulations with hnn-core using MPIBackend and traced it back to a change where the Network object is passed from MPIBackend.simulate() to `mpi_child` via pickling. It can be fixed by passing the Param dict instead of Network object.

Can be reproduced with:
```
while True:
    with MPIBackend(n_procs=8, mpi_cmd='mpiexec'):
        dpls = simulate_dipole(net, n_trials=1, postproc=False)
```

The failure:
```
[Blakes-MacBook-Pro:29089] *** Process received signal ***
[Blakes-MacBook-Pro:29089] Signal: Segmentation fault: 11 (11)
[Blakes-MacBook-Pro:29089] Signal code: Address not mapped (1)
[Blakes-MacBook-Pro:29089] Failing at address: 0x88
[Blakes-MacBook-Pro:29089] [ 0] 0   libsystem_platform.dylib            0x00007fff20384d7d _sigtramp + 29
[Blakes-MacBook-Pro:29089] [ 1] 0   ???                                 0x0000000000000000 0x0 + 0
[Blakes-MacBook-Pro:29089] [ 2] 0   libopen-pal.40.dylib                0x0000000103399a07 opal_libevent2022_event_base_loop + 2087
[Blakes-MacBook-Pro:29089] [ 3] 0   mpiexec                             0x000000010328acf3 orterun + 275
[Blakes-MacBook-Pro:29089] [ 4] 0   libdyld.dylib                       0x00007fff2035b621 start + 1
[Blakes-MacBook-Pro:29089] [ 5] 0   ???                                 0x0000000000000009 0x0 + 9
[Blakes-MacBook-Pro:29089] *** End of error message ***
/Users/blake/miniconda3/envs/hnn-test/lib/python3.7/site-packages/hnn_core/parallel_backends.py:438: UserWarning: Timed out (5s) waiting for end of data after child process stopped
  warn("Timed out (5s) waiting for end of data after child "
```

**Comment by blakecaldwell** [comment]
I realize that hnn-core has changed such that passing a Net object is needed for allowing the full range of customizations to the feeds. However, it makes running optimization unreasonably slow in HNN. HNN will retry failed simulations and event with this issue simulations are eventually successful.

Could you investigate what part of the Network object is causing the seg fault? Any invalid memory address after being passed to the mpi_child process?

**Comment by cjayb** [comment]
I'll try to look into this. We've been bending over backwards to avoid passing around parameter dicts. I'm hoping we could avoid it, it really messes with the code design.

**Comment by jasmainak** [comment]
maybe we need a test that makes sure that pickling and unpickling the `Network` object yields the same object ? I don't see where else this issue could be coming from ...

**Comment by cjayb** [comment]
Hey @blakecaldwell , I'm having troubles reproducing this (`python=3.8.6`, `NEURON=7.8.2`, `openmpi=4.1.0`, `mpi4py=3.0.3`, macOS 10.15.7). See [gist here](https://gist.github.com/cjayb/4c4c2a46229f82ce940f0b58ee5e5675). The 20 iterations I ran completed without error. Note that the network is run for just 70 ms, but with all 3 evoked-like drives.

I'm also at a loss as to what would cause the pickling to fail!

**Comment by blakecaldwell** [comment]
Okay, not terribly surprised. Errors with segmentation faults can be dependent on variability in the hardware, timing (speed), number of cores, etc. Also, pickling works fine. A segmentation fail is invalid memory access. So something is packed into the `Network` object that is valid in the `simulate` in `parallel_backends.py` but is not valid in the child process. This seems plausible given that there have been code changes to Network in the past few months (i.e. more objects being instantiated before passing to `mpi_child.py`). Hoping you have some ideas?

What about stressing things a little more? Run more procs that you have cores... Run longer simulations...

**Comment by blakecaldwell** [comment]
Also if you have committed for changes you suspect could relate to this issue, I'm happy to try them out. I just don't know where to look because I'm not so familiar with the new code.

**Comment by cjayb** [comment]
I increased `n_procs` to 8 (on my quad-core i7 Air), ran the [20 trials with the `default` network](https://gist.github.com/cjayb/4c4c2a46229f82ce940f0b58ee5e5675) (170 ms, three evoked inputs).  No segfaults... I could run more iterations, but your experience is every 3-5 sims, so I'm not sure that'd help. Could you confirm that the above setup causes the segfaults at your end?

The biggest change in `Network` (until #276 gets merged, at least), was the inclusion of methods for drive definition, and dicts for holding event times.

I just ran my script on @ntolley 's #276. Still no segfaults. Passing the `net` to the ranks is insanely slow (>30 sec), but including #287 reduces the time by a factor of 10. That's amazing @blakecaldwell ! @ntolley I recommend a speedy rebase!

**Comment by blakecaldwell** [comment]
Thanks for testing @cjayb. Testing your gist helped me narrow down where the issue might be and realize that I didn't include the key line that made caused the error for me: `net = Network(params)`. The way I was testing is not very useful in practice because there are no feeds. However, HNN GUI optimization also reproduced the issue with different params. Here's the gist that might reproduce it (within 5 simulations for me):
https://gist.github.com/blakecaldwell/b41368ac61cd33b491d403a73f60e32f

Using print statements, I found the error is related to reading in the network object over stdin. It might be related to an increased size of the object. Regardless it has nothing with the `Network` code itself.

The real fix will be what we are discussing in #215. If you would be okay with a temporary hack of mine to make HNN GUI more reliable in the meantime, I may just propose adding a sleep statement. I will follow up with a PR soon.

**Comment by cjayb** [comment]
Still no segfaults after 20 iterations... I don't follow your print-debugging though. Do you mean this line?

```python
events = self.sel.select(timeout=1)
```

If you can make it work with a `sleep`, I'll be more than happy to hit Merge and keep the integration rolling :)
